views/middleware.go

```diff
@@ -14,46 +14,232 @@ import (
 
 // RequiresLogin is a middleware which will be used for each
 // httpHandler to check if there is any active session
-func (v *Views) RequiresLogin(h http.Handler) http.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request) {
-		session, err := v.cookie.Get(r, v.conf.SessionCookieName)
+func (v *Views) RequiresLogin(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		session, err := v.cookie.Get(c.Request(), v.conf.SessionCookieName)
 		if err != nil {
 			log.Println(err)
-			http.Redirect(w, r, "/login", http.StatusFound)
-			return
+			return v.LoginFunc(c)
 		}
-		if !helpers.GetUser(session).Authenticated {
+		user1 := helpers.GetUser(session)
+		user2, err := v.user.GetUser(c.Request().Context(), user1)
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+		if user2.DeletedBy.Valid || !user1.Enabled {
+			session.Values["user"] = &user.User{}
+			session.Options.MaxAge = -1
+			err = session.Save(c.Request(), c.Response())
+			if err != nil {
+				//http.Error(w, err.Error(), http.StatusInternalServerError)
+				return v.errorHandle(c, err)
+			}
+			return c.Redirect(http.StatusFound, "/")
+		}
+		if !user1.Authenticated {
 			// Not authenticated
-			http.Redirect(w, r, "/", http.StatusFound)
-			return
+			return c.Redirect(http.StatusFound, "/")
 		}
-		h.ServeHTTP(w, r)
+		return next(c)
 	}
 }
 
 // RequiresMinimumPermission is a middleware that will
 // ensure that the user has the given permission.
-func (v *Views) RequiresMinimumPermission(h http.Handler, p permissions.Permissions) http.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request) {
-		session, err := v.cookie.Get(r, v.conf.SessionCookieName)
+func (v *Views) RequiresMinimumPermission(next echo.HandlerFunc, p permissions.Permissions) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		session, err := v.cookie.Get(c.Request(), v.conf.SessionCookieName)
+		if err != nil {
+			log.Println(err)
+			http.Error(c.Response(), err.Error(), http.StatusInternalServerError)
+			return v.LoginFunc(c)
+		}
+
+		u := helpers.GetUser(session)
+
+		perms, err := v.user.GetPermissionsForUser(c.Request().Context(), u)
+		if err != nil {
+			log.Println(err)
+			http.Error(c.Response(), err.Error(), http.StatusInternalServerError)
+			return v.LoginFunc(c)
+		}
+
+		acceptedPerms := permission.SufficientPermissionsFor(p)
+
+		for _, perm := range perms {
+			if acceptedPerms[perm.Name] {
+				return next(c)
+			}
+		}
+
+		c.Response().WriteHeader(http.StatusForbidden)
+		return v.Error500(c)
+	}
+}
+
+// RequiresMinimumPermissionMMP is a middleware that will
+// ensure that the user has ManageMembersPermissions.
+func (v *Views) RequiresMinimumPermissionMMP(next echo.HandlerFunc) echo.HandlerFunc {
```

Comment:
NITPICK: Rather than having one for each such permission, would it maybe make sense to have "higher-order middlewares" (for lack of a better word) - e.g.

```go
func RequirePermission(perm Permission) echo.MiddlewareFunc {
  return func(next echo.HandlerFunc) echo.HandlerFunc {
    // ...
  }
}
```

and then use it like

```go
g := echo.Group("...")
g.Use(RequirePermission(permissions.ManageMembersPermission))
```

Reply:
I forgot you can do that and will implement that in a bit

Reply:
Tomorrow problem

Reply:
Implemented